pkg/uploader/provider/kopia.go

```diff
@@ -0,0 +1,226 @@
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/kopia/kopia/snapshot/snapshotfs"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/vmware-tanzu/velero/pkg/uploader"
+	"github.com/vmware-tanzu/velero/pkg/uploader/kopia"
+
+	"github.com/vmware-tanzu/velero/internal/credentials"
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	repokeys "github.com/vmware-tanzu/velero/pkg/repository/keys"
+	"github.com/vmware-tanzu/velero/pkg/repository/udmrepo"
+	"github.com/vmware-tanzu/velero/pkg/repository/udmrepo/service"
+)
+
+//BackupFunc mainly used to make testing more convenient
+var BackupFunc = kopia.Backup
+var RestoreFunc = kopia.Restore
+
+//kopiaProvider recorded info related with kopiaProvider
+//action which means provider handle backup or restore
+type kopiaProvider struct {
+	action         string
+	repoIdentifier string
+	bkRepo         udmrepo.BackupRepo
+	snapshotInfo   *uploader.SnapshotInfo
+	credGetter     *credentials.CredentialGetter
+	log            logrus.FieldLogger
+	uploader       *snapshotfs.Uploader
+	restoreCancel  chan struct{}
+}
+
+//NewKopiaUploaderProvider initialized with open or create a repository
+func NewKopiaUploaderProvider(
+	ctx context.Context,
+	repoIdentifier string,
+	credGetter *credentials.CredentialGetter,
+	repoKeySelector *v1.SecretKeySelector,
+	repoUID string,
+	configFile string,
+	log logrus.FieldLogger,
+	action string,
+) (Provider, error) {
+	kp := &kopiaProvider{
+		repoIdentifier: repoIdentifier,
+		log:            log,
+		action:         action,
+		credGetter:     credGetter,
+	}
+
+	repoOpt, err := udmrepo.NewRepoOptions(
+		udmrepo.WithPassword(kp, ""),
+		udmrepo.WithConfigFile("", repoUID),
+		udmrepo.WithDescription("Initial kopia uploader provider"),
+	)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error to get repo options")
+	}
+
+	repoSvc := service.Create(log)
+	log.WithField("configFile", configFile).Info("Opening backup repo")
+
+	kp.bkRepo, err = repoSvc.Open(ctx, *repoOpt)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to find kopia repository")
+	}
+	return kp, nil
+}
+
+//CheckCancel check context status periodically
+//check if context timeout or cancel
+func (kp *kopiaProvider) CheckCancel(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			if kp.action == uploader.VeleroBackup {
+				kp.uploader.Cancel()
+				kp.log.Infof("Backup is been canceled")
+			} else if kp.restoreCancel != nil {
+				close(kp.restoreCancel)
+				kp.log.Infof("Restore is been canceled")
+			}
+			return
+		default:
+			time.Sleep(time.Second * 10)
+		}
+	}
+}
+
+func (kp *kopiaProvider) Close(ctx context.Context) {
+	kp.bkRepo.Close(ctx)
+}
+
+//RunBackup which will backup specific path and update backup progress in pvb status
+func (kp *kopiaProvider) RunBackup(
+	ctx context.Context,
+	path string,
+	tags map[string]string,
```

Comment:
Looks like `tags` is not currently used, if so, let's remove it

Reply:
RunBackup is a interface, `tags` is been used by resticProvider